### ISSUE TITLE
OnnxStableDiffusionPipeline produces black images when using an FP16 model [DirectML]

### ISSUE THREAD

**Issue by n00mkrad** [huggingface/diffusers#4321]
### Describe the bug

OnnxStableDiffusionPipeline generates an all-black image when loading an FP16 ONNX model.

### Reproduction

- Convert SD 1.5 (Diffusers) to ONNX using `convert_stable_diffusion_checkpoint_to_onnx.py --fp16`
- Run the script below, replace `model_path` with the path to your converted SD 1.5 ONNX model folder

```py
import torch
from diffusers import OnnxStableDiffusionPipeline, DPMSolverMultistepScheduler

model_path = ".../StableDiffusion-v1.5-onnx"
pipe = OnnxStableDiffusionPipeline.from_pretrained(model_path, torch_dtype = torch.float16, provider="DmlExecutionProvider", safety_checker=None) # Load pipeline
pipe.scheduler = DPMSolverMultistepScheduler.from_config(pipe.scheduler.config, use_karras_sigmas = True) # Use DPM++2M Karras
image = pipe(prompt = "nature", negative_prompt = "blurry, low quality, drawing, sketch, painting, cartoon", num_inference_steps = 15, guidance_scale = 7.0).images[0] # Generate
image.save("onnxtest.png", 'PNG')
```

### System Info

Diffusers 0.19.0, Windows 10 22H2, RTX 4090 with driver 536.40, Python 3.10

**Comment by tianleiwu** [comment]
@n00mkrad, try convert and optimize onnx like the following:
```
python convert_stable_diffusion_checkpoint_to_onnx.py --model_path runwayml/stable-diffusion-v1-5  --output_path  ./sd_v1_5_fp32

python -m onnxruntime.transformers.models.stable_diffusion.optimize_pipeline -i ./sd_v1_5_fp32 -o ./sd_v1_5_fp16 --float16
```

For more information, see https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/python/tools/transformers/models/stable_diffusion/README.md

**Comment by sunhs** [comment]
@tianleiwu This is kind of off topic. I convert Unet to onnx and other models remain in PyTorch. I use ORT to integrate the onnx Unet into the pipeline. For StableDiffusionPipeline, the output images of onnx unet and pytorch unet are slightly different. For StableDiffusionControlnetPipeline, this difference is large. I would like to know whether this is expected (given that the repeating denoising steps may amplify the diffs). I can not find other issues that discuss this problem, no matter in `diffusers`, `onnx`, or `onnxruntime`. So if this is expected, I'll give up debugging anymore (which already took a long time). Otherwise maybe I should create a new issue.

FYI, I further convert the onnx unet to TensorRT, and the inference results of ORT and TRT are the same. So the problem shouldn't be related to ORT, but ONNX.

**Comment by tianleiwu** [comment]
@sunhs, there could be slightly difference in cuda kernel implementation (for example layer normalization), which will make slightly different in result. It is expected that the result from ORT, TRT and Torch will not be exactly same, but shall be small enough and no impact on [those end-to-end metrics](https://huggingface.co/docs/diffusers/conceptual/evaluation).

**Comment by sunhs** [comment]
I'm sorry I made some mistakes. The below description is not correct. In fact, even using controlnet, the result seems the same as PyTorch.

> For StableDiffusionPipeline, the output images of onnx unet and pytorch unet are slightly different. For StableDiffusionControlnetPipeline, this difference is large.
